drivers/LargeBlockSR.py

```diff
@@ -0,0 +1,123 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import SR
+from SR import deviceCheck
+import SRCommand
+import EXTSR
+import util
+import os
+
+CAPABILITIES = ["SR_PROBE", "SR_UPDATE", "SR_SUPPORTS_LOCAL_CACHING",
+                "VDI_CREATE", "VDI_DELETE", "VDI_ATTACH", "VDI_DETACH",
+                "VDI_UPDATE", "VDI_CLONE", "VDI_SNAPSHOT", "VDI_RESIZE", "VDI_MIRROR",
+                "VDI_GENERATE_CONFIG",
+                "VDI_RESET_ON_BOOT/2", "ATOMIC_PAUSE", "VDI_CONFIG_CBT",
+                "VDI_ACTIVATE", "VDI_DEACTIVATE", "THIN_PROVISIONING", "VDI_READ_CACHING"]
+
+CONFIGURATION = [['device', 'local device path (required) (e.g. /dev/sda3)']]
+
+DRIVER_INFO = {
+    'name': 'Large Block SR',
+    'description': 'SR plugin which represents disks as VHD files stored on a local EXT4 filesystem, created inside an LVM volume',
+    'vendor': 'Vates',
+    'copyright': '(C) 2024 Vates',
+    'driver_version': '1.0',
+    'required_api_version': '1.0',
+    'capabilities': CAPABILITIES,
+    'configuration': CONFIGURATION
+    }
+
+DRIVER_CONFIG = {"ATTACH_FROM_CONFIG_WITH_TAPDISK": True}
+
+
+class LargeBlockSR(EXTSR.EXTSR):
+    """Emulating 512b drives for EXT storage repository"""
+
+    DRIVER_TYPE = "largeblock"
+    SECTOR_SIZE = 512
+
+    @staticmethod
+    def handles(srtype):
+        return srtype == LargeBlockSR.DRIVER_TYPE
+
+    def attach(self, sr_uuid):
+        self.override_device()
+        super(LargeBlockSR, self).attach(sr_uuid)
+
+    def detach(self, sr_uuid):
+        super(LargeBlockSR, self).detach(sr_uuid)
+        self.remove_device()
+
+    @deviceCheck
+    def create(self, sr_uuid, size):
+        #TODO: Need to check blocksize of the device before accepting to create a SR.
+        self.override_device()
+        super(LargeBlockSR, self).create(sr_uuid, size)
+        self.remove_device()
+
+    def create_loopdev(self, dev, new_path):
+        cmd = ["losetup", "-f", "-v", "--show", "--sector-size", str(self.SECTOR_SIZE), dev]
+        self.loopdev = util.pread2(cmd).rstrip()
+        cmd = ["ln", "-srv", self.loopdev, new_path]
+        util.pread2(cmd)
+
+    def delete_loopdev(self, dev, new_path):
+            cmd = ["rm", "-fv", new_path] # Remove the symlink
+            util.pread2(cmd)
+
+            # The backing file isn't a symlink if given by ID in device-config but the real device
+            dev = os.path.realpath(dev)
+            loopdev = self.get_loopdev_from_device(dev)
+
+            if loopdev != None:
+                cmd = ["losetup", "-d", loopdev] # Remove the loop device
+                util.pread2(cmd)
+
+    def get_loopdev_from_device(self, device):
+        output = util.pread2(["losetup", "--list"]).rstrip().split("\n")
+        for line in output:
+            line = line.split()
+            loopdev = line[0]
+            dev = line[5]
+            if device in dev:
```

Comment:
Not sure, what's the output of losetup? Seems unsafe to handle relative/absolute paths.